### ISSUE TITLE
build: krb-related errors

### ISSUE THREAD

**Issue by linkwk7** [cockroachdb/cockroach#49734]
Hi developers, When i trying build cockroachdb from source, it will fail when build krb.

**To Reproduce**
Follow the instruction in https://wiki.crdb.io/wiki/spaces/CRDB/pages/181338446/Getting+and+building+from+source, trying to build cockroachdb, but it fail when linking kdb5_util

1.mkdir -p $(go env GOPATH)/src/github.com/cockroachdb
2.cd $(go env GOPATH)/src/github.com/cockroachdb
3.git clone https://github.com/cockroachdb/cockroach
4.cd cockroach
5.make buildshort

**Expected behavior**
Build success

**Additional data / screenshots**
Here is part of the build output

making all in kadmin/dbutil...
gcc -L../../lib -Wl,--enable-new-dtags -Wl,-rpath -Wl,/usr/local/lib -g -O2   -o kdb5_util kdb5_util.o kdb5_create.o kadm5_create.o kdb5_destroy.o kdb5_stash.o import_err.o strtok.o dump.o ovload.o kdb5_mkey.o tabdump.o tdumputil.o ../cli/getdate.o -lkadm5srv_mit  -lkdb5 -lkrb5_db2 -lkadm5srv_mit -lgssrpc -lgssapi_krb5 -lgssrpc -lgssapi_krb5 -ldl -pthread  -lkrb5 -lk5crypto -lcom_err -lkrb5support  -lkeyutils -lresolv  -ldl
/usr/bin/ld: ../../lib/libkadm5srv_mit.a(server_kdb.o):/home/rebellion/go/src/github.com/cockroachdb/cockroach/c-deps/krb5/src/lib/kadm5/srv/server_kdb.c:18: multiple definition of `master_keyblock'; kdb5_util.o:(.bss+0x40): first defined here
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:663: kdb5_util] Error 1
make[2]: *** [Makefile:853: all-recurse] Error 1
make[1]: *** [Makefile:1533: all-recurse] Error 1
make: *** [Makefile:776: /home/rebellion/go/native/x86_64-pc-linux-gnu/krb5/lib/libgssapi_krb5.a] Error 2
make: *** Deleting file '/home/rebellion/go/native/x86_64-pc-linux-gnu/krb5/lib/libgssapi_krb5.a'

**Environment:**
 - CockroachDB version 20.1.1, master branch, commit id 021d87644837dfc7e7cd56af386f94b93f41623e
 - Server OS: Arch Linux version 5.6.14-arch1-1 (linux@archlinux) (gcc version 10.1.0 (GCC))
 - Client app: None

**Additional context**

**Comment by lysu** [comment]
Arch Linux User meets the same question too，and comments this line：

https://github.com/cockroachdb/cockroach/blob/ae65915173916d980654206c545ebc829ab8fb51/pkg/ccl/ccl_init.go#L22

can make it pass compile - - if you are not interesting with ccl feature --- gssapi just as me :D

**Comment by dt** [comment]
Hi @lysu ! If you're not interested in having any of the CCL-licensed code linked in, as an alternative to manually commenting out those lines in that package each time you want to build, you can also just use `make buildoss`, which is an alternative `main` package entry point which just doesn't pull in the CCL init imports.

That said, as far as the actual compilation issues, I'll send this over to our C library and kerb experts @knz and @mjibson in the hopes they know what is up, or have an idea who would.

**Comment by linkwk7** [comment]
@lysu @dt Thanks, i use buildoss instead, and it success

**Comment by knz** [comment]
Hi @linkwk7 thanks for your report

If you follow the same steps as you posted initially, but then replace `make buildshort` at the end with:

```
build/builder.sh make buildshort
```

Do you then get a different result?

**Comment by linkwk7** [comment]
Yeah, the output is

Makefile:52: build/defs.mk: No such file or directory
GOPATH set to /go
bash: build/defs.mk.sig: Permission denied
Makefile:812: recipe for target 'build/defs.mk.sig' failed
make: *** [build/defs.mk.sig] Error 1

**Comment by knz** [comment]
@jlinder can you help us qualify this issue further: can you repro the issue reported by the user on a GCE worker?

**Comment by jlinder** [comment]
@knz I spun up a fresh GCE worker to test the build and I was unable to reproduce the problem. I ran these scenarios:

```
cd go/src/github.com/cockroachdb/cockroach/ 
build/builder.sh make buildshort
make buildshort

# checkout the specific commit
git checkout 021d876
make clean
make buildshort
make unsafe-clean
make buildshort
```

Our GCE builder images run Ubuntu so there might be something specific to Arch Linux?

**Comment by knz** [comment]
Thanks James. Your finding suggests that the problem is caused by the user embedding a copy of Kerberos in either their libc or one of the other libraries implicitly linked by the command listed at top. It's not standard to embed Kerberos in this way.

@linkwk7 what we recommend from here is:

1) don't use a root account to download and build crdb

2) always use build/builder.sh as prefix for all commands

On your system I think it's going to be impossible to use 'make' from your shell directly.  Also don't use root.

Can you try again with these instructions and report the results?

**Comment by linkwk7** [comment]
@knz I try build/build.sh make buildshort, and it succeed. Thx

**Comment by knz** [comment]
Perfect, thank you

**Comment by andreimatei** [comment]
I'm running into the same after upgrading to Ubuntu 20.10 (gcc 10.2). Used to work on Ubuntu 20.04.

```
gcc -L../../lib -Wl,--enable-new-dtags -Wl,-rpath -Wl,/usr/local/lib -g -O2   -o kdb5_util kdb5_util.o kdb5_create.o kadm5_create.o kdb5_destroy.o kdb5_stash.o import_err.o strtok.o dump.o ovload.o kdb5_mkey.o tabdump.o tdumputil.o ../cli/getdate.o -lkadm5srv_mit  -lkdb5 -lkrb5_db2 -lkadm5srv_mit -lgssrpc -lgssapi_krb5 -lgssrpc -lgssapi_krb5 -ldl -pthread  -lkrb5 -lk5crypto -lcom_err -lkrb5support  -lresolv  -ldl
/usr/bin/ld: ../../lib/libkadm5srv_mit.a(server_kdb.o):/home/andrei/work/src/github.com/cockroachdb/cockroach/c-deps/krb5/src/lib/kadm5/srv/server_kdb.c:18: multiple definition of `master_keyblock'; kdb5_util.o:/home/andrei/work/src/github.com/cockroachdb/cockroach/c-deps/krb5/src/kadmin/dbutil/kdb5_util.c:109: first defined here
```

This has got to be a gcc/ld version thing, I'm thinking.
@ajwerner as another Ubuntu user, did you upgrade to 20.10? Can you confirm?

**Comment by ajwerner** [comment]
I was on 20.04.

**Comment by knz** [comment]
@jlinder can you accompany andrei / andrew with this. I'm not sure I can help further. Thanks

**Comment by andreimatei** [comment]
@tbg what are you running?

**Comment by andreimatei** [comment]
I've verified that kerberos builds with gcc-9.3 but not with gcc-10.2 (Ubuntu 20.04 vs 20.10). I believe this is because gcc-10 switched to compiling by default with `-fno-common`, which makes the linker error on duplicate definitions. I though the duplicate definitions were always disallowed, but I guess that if they were identical the linker was de-duplicating?
I'll look into patching some makefile, unless someone knows who's more in charge of kerberos...
I'm also quite surprised that I haven't found other reports of this kerberos problem online. Perhaps we've vendored an old version or something?
